### ISSUE TITLE
issueFleetMoveOrder() changes already enqueued route of a fleet

### ISSUE THREAD

**Issue by aschuch247** [freeorion/freeorion#2690]
### Environment

* **FreeOrion Version**: 456a9b64510faf81adfc637b9eb5c106f2d0139e, compiled from source
* **Fetched as**: Compiled from source

### Description

* This is an overview of the situation: ![changed-routing](https://user-images.githubusercontent.com/13714468/71782339-c4d1ac80-2fd0-11ea-8bbb-992bb0e5e7e0.png)
* The marked fleet is at system 879.
* My AI issues the following commands, using `issueFleetMoveOrder()`:
  * Ordering fleet 2072 to move to system 543 with result 1.
  * Ordering fleet 2072 to move to system 777 with result 1.
  * Ordering fleet 2072 to move to system 825 with result 1.
  * Ordering fleet 2072 to move to system 288 with result 1.
  * Ordering fleet 2072 to move to system 903 with result 1.
  * Ordering fleet 2072 to move to system 843 with result 1.
  * Ordering fleet 2072 to move to system 189 with result 1.
  * Ordering fleet 2072 to move to system 909 with result 1.
  * Ordering fleet 2072 to move to system 309 with result 1.
  * Ordering fleet 2072 to move to system 894 with result 1.
* But the server changes the path:
  * Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 2072 to  879 543
  * Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 2072 to  879 543 777
  * Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 2072 to  879 273 825
    * **Here, the path gets changed automatically!**
  * Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 2072 to  879 273 825 288
  * Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 2072 to  879 273 825 288 903
  * Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 2072 to  879 273 825 288 903 843
  * Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 2072 to  879 273 825 288 903 843 189
  * Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 2072 to  879 273 825 288 903 843 189 909
  * Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 2072 to  879 273 825 288 903 843 189 909 309
  * Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 2072 to  879 273 825 288 903 843 189 909 309 894
* As a result, the scout will be destroyed by the space monster at system 273.

### Expected Result

I expect the path to not be altered. Or, if this is an actual feature, I expect another method to exist which has this feature disabled.

### Steps to reproduce

The code used is https://github.com/aschuch247/freeorion/commit/af55a0098368670d19d2b7dc0d00b7f7ef7e0e3e.

The commit identifier might change if I rebase on new upstream. Nevertheless, the commit has the subject 'Enqueue the whole computed path when moving.'

What I do for testing:

* Remove the installation directory of FreeOrion.
* Issue a `make install` in the source code directory.
* Rename `share/freeorion/default/python/AI/FreeOrionElenAI.py` to `share/freeorion/default/python/AI/FreeOrionAI.py`, overwriting the original file.
* Run FreeOrion, using default settings, just reduce the AI players to 1.
* Enable showing identifiers in the FreeOrion options.
* Either build super testers, or just ally with the AI.
* System 273 sometimes has no monster. In this case, retire and start a new game.
* Skip turns until turn 14.
* My AI outputs its actions to `stdout`, the server outputs the log to `~/.local/share/freeorion/AI_1.log`.

**Comment by aschuch247** [comment]
@Vezzra: The instructions for 'new issue' say

> Prefix the title with labels you find appropriate to
> describe the issue like "[Bug]" in case of a bug, "[Feature]" for a feature
> request.

If this is no longer the case, I suggest to edit the template.

**Comment by geoffthemedio** [comment]
That he changed it doesn't mean you shouldn't have followed the instruction to add it.

**Comment by aschuch247** [comment]
@geoffthemedio: I added `[bug]` and it got removed. This is why I wonder.

**Comment by MatGB** [comment]
You can't add tags, only admins can, so the instructions are correct as it
helps the admin pick the tag(s)

**Comment by Vezzra** [comment]
@aschuch247, as I can see, my fellow devs have already answered your question. I hope these answers have cleared the confusion up for you. If not, feel free to ask again. 😃

**Comment by Morlic-fo** [comment]
> I expect the path to not be altered. Or, if this is an actual feature, I expect another method to exist which has this feature disabled.

The function does what a single right-click does for the player: It issues a new move order using the shortest path from the starting system to the target system. This isn't unexpected, probably the interface function was written when there was no support for queueing move orders yet and never updated.

Nowadays,  the ```FleetMoveOrder``` has an optional parameter ```append``` which probably does what you want and should probably work for the AI without further adjustments: https://github.com/freeorion/freeorion/blob/6a3863a4bbead60646df1f3585b7ba2b49cf5e9e/util/Order.h#L200-L201


Relevant lines in the python interface if you want to tackle this: https://github.com/freeorion/freeorion/blob/6a3863a4bbead60646df1f3585b7ba2b49cf5e9e/python/AI/AIWrapper.cpp#L133-L134 https://github.com/freeorion/freeorion/blob/6a3863a4bbead60646df1f3585b7ba2b49cf5e9e/python/AI/AIWrapper.cpp#L205

For reference: https://www.boost.org/doc/libs/1_66_0/libs/python/doc/html/tutorial/tutorial/functions.html#tutorial.functions.default_arguments

**Comment by aschuch247** [comment]
@Morlic-fo: Okay, there is a default parameter in C++, but it does not seem to be exposed to Python.

https://github.com/freeorion/freeorion/blob/master/python/AI/AIWrapper.cpp#L133

I changed the code to read

```
    int IssueFleetMoveOrder(int fleet_id, int destination_id, bool append = false)
    { return Issue<FleetMoveOrder>(fleet_id, destination_id, append); }
```

I can run my AI and the code change is accepted, but it is not working as intended. The following orders are issued, but they fail (result 0).

* Ordering fleet 1017 to move to system 879 with result 1 (reset).
* Ordering fleet 1017 to move to system 543 with result 1 (append).
* Ordering fleet 1017 to move to system 777 with result 0 (append).
* Ordering fleet 1017 to move to system 825 with result 0 (append).
* Ordering fleet 1017 to move to system 288 with result 0 (append).
* Ordering fleet 1017 to move to system 903 with result 0 (append).
* Ordering fleet 1017 to move to system 843 with result 0 (append).
* Ordering fleet 1017 to move to system 189 with result 0 (append).
* Ordering fleet 1017 to move to system 909 with result 0 (append).
* Ordering fleet 1017 to move to system 594 with result 0 (append).

The server logs:

* Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 1017 to  879
* Order.cpp:346 : FleetMoveOrder::ExecuteImpl Setting route of fleet 1017 to  879 543
* Order.cpp:311 : Empire with id 2 ordered a fleet to continue from system with id 879, but the fleet's current route won't lead there, it leads to system 543
* Order.cpp:311 : Empire with id 2 ordered a fleet to continue from system with id 879, but the fleet's current route won't lead there, it leads to system 543
* Order.cpp:311 : Empire with id 2 ordered a fleet to continue from system with id 879, but the fleet's current route won't lead there, it leads to system 543
* Order.cpp:311 : Empire with id 2 ordered a fleet to continue from system with id 879, but the fleet's current route won't lead there, it leads to system 543
* Order.cpp:311 : Empire with id 2 ordered a fleet to continue from system with id 879, but the fleet's current route won't lead there, it leads to system 543
* Order.cpp:311 : Empire with id 2 ordered a fleet to continue from system with id 879, but the fleet's current route won't lead there, it leads to system 543
* Order.cpp:311 : Empire with id 2 ordered a fleet to continue from system with id 879, but the fleet's current route won't lead there, it leads to system 543
* Order.cpp:311 : Empire with id 2 ordered a fleet to continue from system with id 879, but the fleet's current route won't lead there, it leads to system 543

**Comment by aschuch247** [comment]
If I do not change `AIWrapper.cpp`, I get the following error in the AI log file:

```
CommonFramework.cpp:152 : ArgumentError: Python argument types in
    freeOrionAIInterface.issueFleetMoveOrder(int, int, bool)
did not match C++ signature:
    issueFleetMoveOrder(int, int)
```

But maybe some caches are not updated properly? `freeOrionAIInterface.pyi` reads:

```
# Autogenerated do not modify manually!
# This is a type-hinting python stub file, used by python IDEs to provide type hints. For more information
# about stub files, see https://www.python.org/dev/peps/pep-0484/#stub-files
# During execution, the actual freeOrionAIInterface module is made available via
# a C++ Boost-python process as part of the launch.
```

I have no idea when this file actually gets automatically updated, and if this file is even relevant at runtime.

**Comment by Morlic-fo** [comment]
Can you check if https://github.com/freeorion/freeorion/pull/2712 solves your problem (on top of your changes to the AIWrapper)?

**Comment by Morlic-fo** [comment]
> I have no idea ... if this file is even relevant at runtime.

No, it only exists to teach IDEs about the interface functions.

**Comment by Morlic-fo** [comment]
Adding the feature label considering this relates to an extension of the existing AI interface.

**Comment by aschuch247** [comment]
I applied the changes from #2712 to 81f34d2 and it works for me.

**Comment by aschuch247** [comment]
This issue still is not fixed. It misses the exposed API. I created a merge request in #2736.

**Comment by Morlic-fo** [comment]
Assumed to be resolved by #2736. Feel free to report if further problems occur.
